### ISSUE TITLE
Installation order seems to matter

### ISSUE THREAD

**Issue by gbrener** [conda-forge/gdal-feedstock#170]
Hello, I'm experiencing an issue with installing a conda package for a C++ binary that depends on `gdal`. I have an _environment.yml_ file that looks something like this:

```
name: some_env
channels:
  - conda-forge
  - defaults
dependencies:
  - gdal
  - cpp_conda_pkg
```

After a `conda env create && source activate some_env`, executing a Python wrapper for cpp_conda_pkg gives me the following error at runtime:

```
ImportError: /home/<user_name>/miniconda/envs/<env_name>/lib/python3.5/site-packages/osgeo/_gdal.cpython-35m-x86_64-linux-gnu.so: undefined symbol: VSISupportsSparseFiles
```

The error no longer occurs when I re-install `gdal` with the following command:
```
conda install -c conda-forge -f gdal
```

Is this more likely to be a problem with cpp_conda_pkg, or this gdal feedstock? Advice/help in addressing this issue would be much appreciated.

**Comment by ocefpaf** [comment]
@gbrener can you send the results of `conda info` and `conda list` from before and after you force it to re-install?

**Comment by ocefpaf** [comment]
@gbrener I need your `conda info` and `conda list` results otherwise I cannot debug this. I cannot reproduce it locally with your env file. Which means you may have an outdated version of something or it is specific to OS X.

**Comment by gbrener** [comment]
Ok, not a problem @ocefpaf. Below are `conda info` and `conda list` results for both OSX and Linux:

## OSX
`conda info` and `conda list` are the same before and after re-installing `gdal`:

OSX _conda info_:
```
Current conda install:

               platform : osx-64
          conda version : 4.3.24
       conda is private : False
      conda-env version : 4.3.24
    conda-build version : 2.1.17
         python version : 3.6.2.final.0
       requests version : 2.14.2
       root environment : /Users/gbrener/miniconda  (writable)
    default environment : /Users/gbrener/miniconda/envs/some_env
       envs directories : /Users/gbrener/miniconda/envs
                          /Users/gbrener/.conda/envs
          package cache : /Users/gbrener/miniconda/pkgs
                          /Users/gbrener/.conda/pkgs
           channel URLs : https://repo.continuum.io/pkgs/free/osx-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/osx-64
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/osx-64
                          https://repo.continuum.io/pkgs/pro/noarch
            config file : None
             netrc file : /Users/gbrener/.netrc
           offline mode : False
             user-agent : conda/4.3.24 requests/2.14.2 CPython/3.6.2 Darwin/16.5.0 OSX/10.12.4
                UID:GID : -
```

OSX _conda list_:
```
# packages in environment at /Users/gbrener/miniconda/envs/some_env:
#
affine                    2.1.0                    py35_0    conda-forge
attrs                     17.2.0                   py35_0    conda-forge
boost                     1.64.0                   py35_3    conda-forge
boost-cpp                 1.64.0                        1    conda-forge
boto3                     1.4.4                    py35_0    conda-forge
botocore                  1.5.92                   py35_0    conda-forge
bzip2                     1.0.6                         1    conda-forge
ca-certificates           2017.7.27.1                   0    conda-forge
cairo                     1.14.6                        4    conda-forge
certifi                   2017.7.27.1              py35_0    conda-forge
click                     6.7                      py35_0    conda-forge
click-plugins             1.0.3                    py35_0    conda-forge
cligj                     0.4.0                    py35_0    conda-forge
curl                      7.54.1                        0    conda-forge
docutils                  0.13.1                   py35_0    conda-forge
expat                     2.1.0                         3    conda-forge
fontconfig                2.12.1                        4    conda-forge
freetype                  2.7                           1    conda-forge
freexl                    1.0.2                         2    conda-forge
gdal                      2.2.0               np112py35_2    conda-forge
geos                      3.5.1                         1    conda-forge
gettext                   0.19.8.1                      0    conda-forge
giflib                    5.1.4                         0    conda-forge
glib                      2.51.4                        0    conda-forge
hdf4                      4.2.12                        0    conda-forge
hdf5                      1.8.18                        1    conda-forge
hypothesis                3.17.0                   py35_0    conda-forge
icu                       58.1                          1    conda-forge
jmespath                  0.9.3                    py35_0    conda-forge
jpeg                      9b                            0    conda-forge
json-c                    0.12.1                        0    conda-forge
kealib                    1.4.7                         2    conda-forge
krb5                      1.14.2                        0    conda-forge
libdap4                   3.18.3                        2    conda-forge
libffi                    3.2.1                         3    conda-forge
libgdal                   2.1.4                         2    conda-forge
libgfortran               3.0.0                         0    conda-forge
libiconv                  1.14                          4    conda-forge
libkml                    1.3.0                         1    conda-forge
libnetcdf                 4.4.1.1                       6    conda-forge
libpng                    1.6.28                        1    conda-forge
libpq                     9.6.3                         0    conda-forge
libspatialite             4.3.0a                       15    conda-forge
libssh2                   1.8.0                         1    conda-forge
libtiff                   4.0.6                         7    conda-forge
libxml2                   2.9.4                         4    conda-forge
mkl                       2017.0.3                      0
ncurses                   5.9                          10    conda-forge
numpy                     1.12.1                   py35_0
openjpeg                  2.1.2                         2    conda-forge
openssl                   1.0.2l                        0    conda-forge
pcre                      8.39                          0    conda-forge
pip                       9.0.1                    py35_0    conda-forge
pixman                    0.34.0                        0    conda-forge
poppler                   0.52.0                        2    conda-forge
poppler-data              0.4.7                         0    conda-forge
proj4                     4.9.3                         4    conda-forge
pympler                   0.5                      py35_0    conda-forge
pyparsing                 2.2.0                    py35_0    conda-forge
python                    3.5.4                         0    conda-forge
python-dateutil           2.6.1                    py35_0    conda-forge
rasterio                  0.36.0                   py35_0    conda-forge
readline                  6.2                           0    conda-forge
s3transfer                0.1.10                   py35_1    conda-forge
setuptools                36.2.2                   py35_0    conda-forge
six                       1.10.0                   py35_1    conda-forge
snuggs                    1.4.1                    py35_0    conda-forge
sqlite                    3.13.0                        1    conda-forge
tk                        8.5.19                        2    conda-forge
wheel                     0.29.0                   py35_0    conda-forge
xerces-c                  3.1.4                         3    conda-forge
xz                        5.2.3                         0    conda-forge
zlib                      1.2.8                         3    conda-forge
zope                      1.0                      py35_0
zope.interface            4.4.2                    py35_0
```

## Ubuntu
On Linux, `conda info` and `conda list` output also does not change after re-installing gdal:

Linux _conda info_:
```
Current conda install:

               platform : linux-64
          conda version : 4.3.21
       conda is private : False
      conda-env version : 4.3.21
    conda-build version : not installed
         python version : 3.6.1.final.0
       requests version : 2.14.2
       root environment : /home/gbrener/miniconda  (writable)
    default environment : /home/gbrener/miniconda/envs/some_env
       envs directories : /home/gbrener/miniconda/envs
                          /home/gbrener/.conda/envs
          package cache : /home/gbrener/miniconda/pkgs
                          /home/gbrener/.conda/pkgs
           channel URLs : https://repo.continuum.io/pkgs/free/linux-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/linux-64
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/linux-64
                          https://repo.continuum.io/pkgs/pro/noarch
            config file : None
             netrc file : None
           offline mode : False
             user-agent : conda/4.3.21 requests/2.14.2 CPython/3.6.1 Linux/4.9.36-moby debian/jessie/
                UID:GID : -
```

Linux _conda list_:
```
# packages in environment at /home/gbrener/miniconda/envs/some_env:
#
affine                    2.1.0                    py35_0    conda-forge
attrs                     17.2.0                   py35_0    conda-forge
boost                     1.64.0                   py35_3    conda-forge
boost-cpp                 1.64.0                        1    conda-forge
boto3                     1.4.4                    py35_0    conda-forge
botocore                  1.5.92                   py35_0    conda-forge
bzip2                     1.0.6                         1    conda-forge
ca-certificates           2017.7.27.1                   0    conda-forge
cairo                     1.14.6                        4    conda-forge
certifi                   2017.7.27.1              py35_0    conda-forge
click                     6.7                      py35_0    conda-forge
click-plugins             1.0.3                    py35_0    conda-forge
cligj                     0.4.0                    py35_0    conda-forge
curl                      7.54.1                        0    conda-forge
docutils                  0.13.1                   py35_0    conda-forge
expat                     2.1.0                         3    conda-forge
fontconfig                2.12.1                        4    conda-forge
freetype                  2.7                           1    conda-forge
freexl                    1.0.2                         2    conda-forge
gdal                      2.2.0               np112py35_2    conda-forge
geos                      3.5.1                         1    conda-forge
gettext                   0.19.7                        1    conda-forge
giflib                    5.1.4                         0    conda-forge
glib                      2.51.4                        0    conda-forge
hdf4                      4.2.12                        0    conda-forge
hdf5                      1.8.18                        1    conda-forge
hypothesis                3.17.0                   py35_0    conda-forge
icu                       58.1                          1    conda-forge
jmespath                  0.9.3                    py35_0    conda-forge
jpeg                      9b                            0    conda-forge
json-c                    0.12.1                        0    conda-forge
kealib                    1.4.7                         2    conda-forge
krb5                      1.14.2                        0    conda-forge
libdap4                   3.18.3                        2    conda-forge
libffi                    3.2.1                         3    conda-forge
libgdal                   2.1.4                         2    conda-forge
libgfortran               3.0.0                         1
libiconv                  1.14                          4    conda-forge
libkml                    1.3.0                         1    conda-forge
libnetcdf                 4.4.1.1                       6    conda-forge
libpng                    1.6.28                        1    conda-forge
libpq                     9.6.3                         0    conda-forge
libspatialite             4.3.0a                       15    conda-forge
libssh2                   1.8.0                         1    conda-forge
libtiff                   4.0.6                         7    conda-forge
libxml2                   2.9.4                         4    conda-forge
mkl                       2017.0.3                      0
ncurses                   5.9                          10    conda-forge
numpy                     1.12.1                   py35_0
openjpeg                  2.1.2                         3    conda-forge
openssl                   1.0.2l                        0    conda-forge
pcre                      8.39                          0    conda-forge
pip                       9.0.1                    py35_0    conda-forge
pixman                    0.34.0                        0    conda-forge
poppler                   0.52.0                        2    conda-forge
poppler-data              0.4.7                         0    conda-forge
proj4                     4.9.3                         4    conda-forge
pympler                   0.5                      py35_0    conda-forge
pyparsing                 2.2.0                    py35_0    conda-forge
python                    3.5.4                         0    conda-forge
python-dateutil           2.6.1                    py35_0    conda-forge
rasterio                  0.36.0                   py35_0    conda-forge
readline                  6.2                           0    conda-forge
s3transfer                0.1.10                   py35_1    conda-forge
setuptools                36.2.2                   py35_0    conda-forge
six                       1.10.0                   py35_1    conda-forge
snuggs                    1.4.1                    py35_0    conda-forge
sqlite                    3.13.0                        1    conda-forge
tk                        8.5.19                        2    conda-forge
util-linux                2.21                          0
wheel                     0.29.0                   py35_0    conda-forge
xerces-c                  3.1.4                         3    conda-forge
xz                        5.2.3                         0    conda-forge
zlib                      1.2.8                         3    conda-forge
zope                      1.0                      py35_0
zope.interface            4.4.2                    py35_0
```

Thanks again for looking into this.

**Comment by ocefpaf** [comment]
I need find if we have some old unpinned (or bad pinned) package that is causing this and remove it from the channel. In theory, if conda respected the higher build number, this should not happen.

Here is a temporary workaround until I fix this.

```yaml
name: some_env
channels:
  - conda-forge
  - defaults
dependencies:
  - python=3.5
  - gdal=2.1.*
  - rasterio
```

By forcing `gdal 2.1.*` in the env you should get all packages consistent with that version of `gdal`.

**Comment by ocefpaf** [comment]
`rasterio` is properly pinned to `libgdal 2.1.*`, that is why you get that version of `libgdal`, but because you specify `gdal` (the python wrapper) we end up with the latest version of `gdal` that requires a different version of the `libgdal` installed. IMO conda is fundamentally broken when it allows something like that to happen.

Unfortunately conda has is no advanced concept of incompatibility of packages like RPMs.
The solutions are to "fool" conda to do the right thing like the env I sent you above, or changing `rasterio` to use `gdal 2.1.*`, which will pull the right `libgdal`, instead of `libgdal` :unamused: 

I recommend you do the former for now but I'll look into using the latter as a long term solution.

**Comment by msarahan** [comment]
FWIW, I don't think there's a fundamental problem with conda.  Version constraints can capture the incompatibility here.  The crappy thing is that without conda-build 3's pin_compatible, it's much harder to keep track of.  @ocefpaf if I can help you understand how to modify these recipes so that conda-build 3 helps capture compatibility bounds here, let me know.  Might be good to have a quick screen-sharing session.

**Comment by ocefpaf** [comment]
> FWIW, I don't think there's a fundamental problem with conda.  Version constraints can capture the incompatibility here.  The crappy thing is that without conda-build 3's pin_compatible, it's much harder to keep track of.

I still believe that how conda treats build numbers is "broken," but we don't need to debate that.
And yes, `conda-build 3` seems the way to go.

>  @ocefpaf if I can help you understand how to modify these recipes so that conda-build 3 helps capture compatibility bounds here, let me know.  Might be good to have a quick screen-sharing session.

Sure let us do that! Schedule a time? Or do you want to try a PR to `libgdal` as an example that I can follow up later?

**Comment by msarahan** [comment]
I will submit a PR, and then let's schedule a time to go over it on a call.  It'll be easier to explain that way.  I'll also comment it a lot.  We can decide whether to keep the comments or not later.

**Comment by ocefpaf** [comment]
Awesome! Sound like a plan. Let's us that experiment to write some docs on how to move to conda-build 3 for conda-forge. (I'll write you review :wink:)

**Comment by davidh-ssec** [comment]
FYI I had this same issue when building a package that depends on gdal, rasterio, and python 3.6. I did the recommended workaround, but got an error about unsatisfiable dependencies and listing only gdal. I then switched it to just `gdal ==2.1`, but that failed since that exact version is only available on python 3.5. I did get it to work by doing `gdal >=2.1,<2.2`.
